### DISCUSSION TITLE
How to get the vSphere 6.7 version API

### DISCUSSION THREAD

**Discussion by pgeorgie1** [vmware/govmomi#3549] [Community Help]
I would like to ask how can I find the vSphere 6.7 version API in the current repo - is there any specific branch, or change set that is delivering the vSphere 6.7 APIs?

**Comment by karaatanassov** [comment]
Can you share a bit more about the use case for 6.7 API. Do you want the various data objects to be pruned to their 6.7 version or do you simply need govmomi to talk to vCenter 6.7. 

The latter is attainable with the latest govmomi release by simply pointing the library at vCenter or ESX version 6.7. It will be incumbent upon the application to send requests that a 6.7 server understands. This means different things for different APIs. For example with virtual machines the hardware version and consequently the types of supported devices would be limited to what vCenter and ESX 6.7 understand. For other APIs this would mean not sending data types and fields added after 6.7. Note that most operations and attributes in the API have [documentation about the time they were shipped](https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U3/html/ReferenceGuides/vim.vm.ConfigSpec.html) i.e. these "Since vSphere API Release 6.7" notes. This would be recommended way to work as you get all the upgrades to govmomi including compatibility with newer golang features. You will also be able to leverage features of new vCenter or ESX APIs quickly if and when a need arises. 

Alternatively I suppose you can go back to govmomi versions from mid/late 2018 or early 2019 e.g. v 0.21 or so. This would also downgrade the overall library, utility code, compatibility with golang, security fixes etc. So such trip in time would not be nice.

A third option would be to get the XSD/WSDL API definitions from [vCenter 6.7 Management SDK](https://developer.broadcom.com/sdks/vsphere-management-sdk/6.7) and generate the data types and functions from that model. This may void compatibility with some of the utility code in govmomi that will not be able to compile with the much older API specification.

**Comment by pgeorgie1** [comment]
Hello Kiril,

My use case is very simple - I need to extract performance metric (provided by the performance manager object) from heterogeneous cluster of vCenter servers - from 6.7 to 8.0.2 versions.

I have already tested that I am able to collect the performance metric from all version of the deployed vCenter servers by using java vSphere APIs (vmomi) for version 6.7 - a backward compatibility setup where single lowest API is managing multiple major product versions.  
Basically, I am looking to recreate the same use case in golang and I am trying to find a suitable govmomi version (golang vSphere API dependency).

My question is - can I  achieve the following simple and very generic use case if i use the latest golang vSphere APIs(govmomi) against vSphere version: 6.7, 7.0.3 and 8.0.2:

 - lookup for ESX managed objects on a vCenter instance
 - collecting a predefined set of performance metric for the found ESX managed object by using the Performance Manager service

Thanks in advance,
Petar

**Reply by dougm** [reply]
In general, govmomi's API bindings should work with the same versions annotated in the sdk docs. What you describe here would likely work with versions as far back as 5.0 with govmomi. However, govmomi project maintainers are currently only concerned with versions 7.0 and higher: https://github.com/vmware/govmomi/blob/main/README.md#compatibility
